### PR TITLE
[Core] Multi-tenancy: Job isolation & implement per job config (except for env variables)

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

* Implement per job config in raylet and Java/Python worker. (Setting env variables for worker processes will be implemented in a later PR.)
* Enforce job isolation. Each worker process is bound to a job and can only execute tasks of that job.

### Checklist

- [x] No shared worker pool.
- [x] Job isolation by binding each worker process to a job.
- [x] No num initial workers by default.
- [x] Fix the backward compatibility issue for the benchmark case.
- [ ] Capping the total number of workers in a node.
- [x] Add a feature flag. 

## Related issue number

#9373 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/job_configs.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        num_java_workers_per_process (int): The number of java workers per
+            worker process.
+        jvm_options (str[]): The jvm options for java workers of the job.
+    """
+
+    def __init__(
+            self,
+            num_initial_python_workers=-1,
+            num_initial_java_workers=-1,
+            worker_env=dict(),
```
Comment (rkooo567) [inline python/ray/job_configs.py:25]: We should avoid using mutable values as default. https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

File: python/ray/job_configs.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            num_java_workers_per_process=10,
+            jvm_options=[],
```
Comment (rkooo567) [inline python/ray/job_configs.py:27]: Same here.

File: python/ray/job_configs.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import ray
+
+
+class JobConfigs:
```
Comment (ericl) [inline python/ray/job_configs.py:4]:
```suggestion
class JobConfig:
```

Comment (ericl) [reply python/ray/job_configs.py:4]: Config is plural, you don't need the extra s. That would imply configs for multiple jobs.

File: python/ray/worker.py [hunk 4]
```diff
@@ -486,6 +487,7 @@ def init(address=None,
          dashboard_host="localhost",
          dashboard_port=ray_constants.DEFAULT_DASHBOARD_PORT,
          job_id=None,
+         job_configs=None,
```
Comment (ericl) [inline python/ray/worker.py:490]:
```suggestion
         job_config=None,
```

Comment (ericl) [reply python/ray/worker.py:490]: Please rename this, here and elsewhere.

File: src/ray/raylet/node_manager.cc [hunk 5]
```diff
@@ -160,11 +160,19 @@ NodeManager::NodeManager(boost::asio::io_service &io_service,
...
-          config.min_worker_port, config.max_worker_port, gcs_client_,
-          config.worker_commands, config.raylet_config,
+          io_service, config.adaptive_num_initial_workers,
+          config.maximum_startup_concurrency, config.min_worker_port,
+          config.max_worker_port, gcs_client_, config.worker_commands,
+          config.raylet_config,
           /*starting_worker_timeout_callback=*/
-          [this]() { this->DispatchTasks(this->local_queues_.GetReadyTasksByClass()); }),
+          [this]() { this->DispatchTasks(this->local_queues_.GetReadyTasksByClass()); },
+          [this](const JobID &job_id) -> const rpc::JobConfigs * {
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc:169]:
```suggestion
          [this](const JobID &job_id) -> boost::optional<rpc::JobConfig> {
```

File: src/ray/raylet/node_manager.h [hunk 6]
```diff
@@ -64,8 +64,9 @@ struct NodeManagerConfig {
   /// The highest port number that workers started will bind on.
   /// If this is not set to 0, min_worker_port must also not be set to 0.
   int max_worker_port;
-  /// The initial number of workers to create.
-  int num_initial_workers;
+  /// Number of initial workers to start per job if num_initial_***_workers
+  /// is not specified in the job config.
+  uint32_t adaptive_num_initial_workers;
```
Comment (ericl) [inline src/ray/raylet/node_manager.h:69]:
```suggestion
  uint32_t default_num_initial_workers;
```

File: src/ray/raylet/node_manager.h [hunk 7]
```diff
@@ -720,6 +721,9 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
   /// The resources (and specific resource IDs) that are currently available.
   ResourceIdSet local_available_resources_;
   std::unordered_map<ClientID, SchedulingResources> cluster_resource_map_;
+
+  absl::flat_hash_map<JobID, rpc::JobTableData> job_info_cache_;
```
Comment (ericl) [inline src/ray/raylet/node_manager.h:725]: Please add a comment here about how this is populated and cleaned up.

File: src/ray/raylet/worker_pool.cc [hunk 8]
```diff
@@ -343,19 +349,59 @@ Status WorkerPool::RegisterWorker(const std::shared_ptr<Worker> &worker, pid_t p
...
   RAY_RETURN_NOT_OK(GetNextFreePort(port));
   driver->SetAssignedPort(*port);
   auto &state = GetStateForLanguage(driver->GetLanguage());
   state.registered_drivers.insert(std::move(driver));
+  driver->AssignJobId(job_id);
+
   return Status::OK();
 }
 
+void WorkerPool::StartInitialWorkersForJob(const JobID &job_id) {
```
Comment (ericl) [inline src/ray/raylet/worker_pool.cc:376]: So say I have N cores, and launch M jobs. Does this mean I end up with N*M workers by default? That seems likely to break existing workloads following this pattern.

**Review by ericl — CHANGES_REQUESTED** [review]
Hey @kfstorm , great pr. A couple questions
- Previously the size of the worker pool is normally bounded by num_cpus. Is it possible to enforce this bound still to avoid resource exhaustion?
- How do we clean up workers when the job exits?

I'm also wondering if it's a good idea to feature flag this new handling of workers. Currently the PR is written in a way that seems like its hard to revert to the old behaviour.

**Comment by ericl** [comment]
Btw, another issue is that the current implementation breaks warm-start of Ray. For example, the following script
```
import ray
import time

@ray.remote
def f():
    return

ray.init()

start = time.time()
ray.get(f.remote())
print(time.time() - start)
```

Will print out ~`0.0005` before this PR, and ~`0.66` after this PR. This is since the initial workers aren't created by the time init returns.

This is actually a quite serious since (naive) users benchmark Ray without warming it up, so we end up looking terribly slow for a pretty dumb reason. It has happened quite a few times.

I'm not sure about the best resolution for this. One option is to preserve the existing worker pool, and only allocate dedicated workers for jobs with job_config set explicitly. This would also be the best for backwards compatibility. Or, we can fix warmup for the job that calls ray.init().

**Comment by kfstorm** [comment]
Hi @ericl,

> * Previously the size of the worker pool is normally bounded by num_cpus. Is it possible to enforce this bound still to avoid resource exhaustion?

AFAIK, previously only the *initial* size of worker pool is set to `num_cpus`. Users can still spawn a lot of tasks to create more workers. I'm not sure what do you mean about "resource exhaustion".

> * How do we clean up workers when the job exits?

Workers of a job are killed once the job finishes. See https://github.com/ray-project/ray/blob/9318e76b81d7d70ffce948532c2b816676989193/src/ray/raylet/node_manager.cc#L337

> I'm also wondering if it's a good idea to feature flag this new handling of workers. Currently the PR is written in a way that seems like its hard to revert to the old behaviour.

Sure. I'll add a flag and set it make it disabled by default.

File: src/ray/raylet/worker_pool.cc (hunk 8 above)
Comment (kfstorm) [reply src/ray/raylet/worker_pool.cc:376]: You are right. Previously a Ray cluster contains only one job, so M == 1 here. not a problem for existing workloads if we still want to run workloads in standalone clusters.

This design is for backward compatibility. 1) We want to start initial workers when the driver is registered to Raylet. 2) a lot of Python tests rely on initial workers.

**Comment by ericl** [comment]
To be clear, spawning workers is not free, each worker consumes memory,
creates threads, and so on which can hit system limits. It is highly
desirable to bound the number of workers to not much more than the number
of CPUs.

I think one reasonable way to do this is to keep a default pool of shared
workers as before. Then, for isolated jobs, launch workers on demand only
(remove the initial size option). When one of these isolated workers is
launched, destroy an idle worker to keep the number of workers under
control. How does this sound?

**Comment by kfstorm** [comment]
@ericl I'm aware that by default spawning initial workers for each job is heavy and needs to be changed. And I like the idea that if a user explicitly sets the job config, we can choose to not start initial workers for this job at all. But I have some concerns,

1. Shared workers make the system complicated.
    * First of all, we need to track these shared workers in a standalone pool, and add additional logic in `WorkerPool::PopWorker`.
    * Besides, in our internal repo, We've made some changes beyond this PR and they are not compatible with shared workers. e.g. We've added an env variable `RAY_JOB_ID` to every worker process (so some code blocks in this PR like [1)](https://github.com/alipay/ray/blob/a730c0d39991c7f0eb1fb67827b6bfa9950f6e2b/src/ray/raylet/worker.cc#L109-L118) [2)](https://github.com/alipay/ray/blob/a730c0d39991c7f0eb1fb67827b6bfa9950f6e2b/src/ray/raylet/worker_pool.cc#L245-L248) [3)](https://github.com/alipay/ray/blob/a730c0d39991c7f0eb1fb67827b6bfa9950f6e2b/src/ray/raylet/worker_pool.cc#L355-L359) are already removed in our internal repo). This is also a new way to pass some determined information from Raylet to workers without adding new parameters to worker command, and these env variables can be read by core worker directly, which are beneficial for both Java and Python workers. Potentially, we could add more env variables like node ID, node manager port, assigned worker port, .etc. As long as these parameters are only needed by core worker, we don't even need to update Java/Python code to parse and pass these parameters.

2. For keeping the number of workers under control, it's not a problem introduced by this PR, right? For example, N jobs spawned X workers each because the initial workloads of these jobs are heavy, but then the workloads are not heavy anymore, so a lot of workers are idle but they can't be used by new jobs and they occupy a lot of memory and other resources. Before we have job isolation, let's consider that these jobs are merged into one big job, the idle workers are able to be used by new workflows. But after this PR, new jobs may not be able to run smoothly because resources on nodes are limited and we can't reuse existing workers because they belong to existing jobs. If this is what you are worried about, maybe we can terminate idle workers after a certain time period. BTW, when a job finishes (when a driver disconnects from Raylet), all workers of this job will be killed by Raylet.

I understand that the first impression from users are important. I'm not saying that we absolutely can't keep shared workers. I just want to share my thinkings before we make a decision.

File: python/ray/job_configs.py (hunk 1 above)
Comment (kfstorm) [reply python/ray/job_configs.py]: Thanks! Done.

File: python/ray/job_configs.py (hunk 2 above)
Comment (kfstorm) [reply python/ray/job_configs.py]: Thanks! Done.

File: python/ray/job_configs.py (hunk 3 above)
Comment (kfstorm) [reply python/ray/job_configs.py]: Good point! Done.

File: python/ray/worker.py (hunk 4 above)
Comment (kfstorm) [reply python/ray/worker.py]: Done.

File: src/ray/raylet/node_manager.cc (hunk 5 above)
Comment (kfstorm) [reply src/ray/raylet/node_manager.cc]: Done.

File: src/ray/raylet/node_manager.h (hunk 6 above)
Comment (kfstorm) [reply src/ray/raylet/node_manager.h]: Done.

File: src/ray/raylet/node_manager.h (hunk 7 above)
Comment (kfstorm) [reply src/ray/raylet/node_manager.h]: Done.

File: src/ray/raylet/worker_pool.cc (hunk 8 above)
Comment (raulchen) [reply src/ray/raylet/worker_pool.cc:376]: What about making "num_initial_workers_per_job" configurable? This is a cluster-level config. For power users who want to run multiple jobs per cluster, the cluster admin can set this to 0. And the job submiters don't need to care about this.

Comment (raulchen) [reply src/ray/raylet/worker_pool.cc:376]: But we still need to deal with the test cases unfortunately.

**Comment by ericl** [comment]
> For keeping the number of workers under control, it's not a problem introduced by this PR, right? 

I think there is a bit of a misconception here, it was not a problem before since there was only ever 1 pool of workers (of fixed size), and the workers are shared among the jobs. The scheduler does a bit of fair queueing to split up the workers between jobs/task types fairly.

Perhaps this is not a common use case of Ray though, and it is OK to add extra worker startup latency for the sake of simplifying the worker pool handling as you mention.

> If this is what you are worried about, maybe we can terminate idle workers after a certain time period.

I would prefer to immediately kill idle workers if the number of total workers exceeds num_cpus, to avoid resource exhaustion that brings down the entire node. This would be similar to current behavior with the shared worker pool.

> I understand that the first impression from users are important. I'm not saying that we absolutely can't keep shared workers. I just want to share my thinkings before we make a decision.

Perhaps we can special case this to avoid needing shared workers. We just need to make sure that when you do a `ray.init()` or `ray.init(address=)`, that it launches at least 1 worker and waits for that worker to start before init returns. That way, you pay zero latency for most common benchmarks. Actually I'm not entirely sure why the current PR breaks this behavior.

So to summarize my updated thinking:
- I agree it is OK to not have a shared pool (though, this would be breaking backwards compatibility in the case of multiple jobs). It would still be nice to have a shared pool available via feature flag (if possible).
- We should try to mask initial task latency by waiting for at least 1 worker.
- We should remove the initial_num_workers parameter.
- We should hard cap the overall pool size to num_cpus (maybe LRU / fair kill idle workers).

Any thoughts? cc @edoakes @stephanie-wang

File: src/ray/raylet/worker_pool.cc (hunk 8 above)
Comment (ericl) [reply src/ray/raylet/worker_pool.cc:376]: Can we fix the Python tests? I can't imagine there are that many.

**Comment by kfstorm** [comment]
> Perhaps we can special case this to avoid needing shared workers. We just need to make sure that when you do a ray.init() or ray.init(address=), that it launches at least 1 worker and waits for that worker to start before init returns. That way, you pay zero latency for most common benchmarks. Actually I'm not entirely sure why the current PR breaks this behavior.

I think I've found the root cause of the breaking. This is a demonstration of timeline comparison:

## Current master

* 00:00:00.000 Driver invokes `ray.init()`.
* 00:00:00.010 Raylet starts.
* 00:00:00.015 Raylet creates the worker pool with initial workers.
* 00:00:00.015 Initial workers spawned.
* 00:00:00.030 Driver tries to connect to Raylet for the first time. (Failure)
* 00:00:00.040 Raylet finishes initialization.
* 00:00:01.015 The first initial worker connects to Raylet.
* 00:00:01.030 Driver tries to connect to Raylet for the second time. (Success)
* 00:00:01.035 `ray.init()` finishes.
* 00:00:01.035 Driver starts benchmark.
* 00:00:01.035 Driver submits a task.
* 00:00:01.040 Worker executes the task.
* 00:00:01.045 Driver gets the task result.
* 00:00:01.045 Driver stops benchmark. The benchmark result is 0.1s.

## This PR

* 00:00:00.000 Driver invokes `ray.init()`.
* 00:00:00.010 Raylet starts.
* 00:00:00.015 Raylet creates an empty worker pool.
* 00:00:00.030 Driver tries to connect to Raylet for the first time. (Failure)
* 00:00:00.040 Raylet finishes initialization.
* 00:00:01.030 Driver tries to connect to Raylet for the second time. (Success)
* 00:00:01.030 Initial workers of this job spawned because Raylet found a new job.
* 00:00:01.035 `ray.init()` finishes.
* 00:00:01.035 Driver starts benchmark.
* 00:00:01.035 Driver submits a task.
* 00:00:02.015 The first initial worker connects to Raylet.
* 00:00:02.020 Worker executes the task.
* 00:00:02.025 Driver gets the task result.
* 00:00:02.025 Driver stops benchmark. The benchmark result is 0.9s.

The key is, A Python worker typically takes about 1s to be registered to Raylet (I believe it's caused by imports of ray modules). In the master branch, A driver takes about 1s to connects to Raylet, which conceals the problem. In this PR, initial workers are spawned after the driver is connected to Raylet, so the benchmark result is worse than the master branch.

If I modify the code to set the raylet client retry interval to a smaller value like 10ms, both the master branch and this PR shows the similar benchmark result of 0.9s.

**Comment by edoakes** [comment]
@ericl is backwards compatibility a significant issue? If I understand correctly, this would only come up if users' application behavior relies on the fact that multiple jobs share the same worker processes (e.g., they share some global variables). That seems pretty niche to me.

Completely agree that we need to vainly optimize cold-start latency, this has shown up in external blogs and issues before. I'm actually a bit worried that even with only waiting for one worker this will still cause problems, e.g., if a user measures something like `ray.get([f.remote() for _ in range 10])` where `f` isn't a complete no-op. We could probably mostly address this by just continuing to start `num_cpus` workers when in "local cluster mode" (`ray.init()`).

Capping idle workers also seems like a good idea, users already complain about idle workers consuming memory sometimes.

**Comment by ericl** [comment]
For backwards compat, the main issue I see is the added latency if you're
launching many small jobs. It would go from a few millis to a couple
seconds per job. However, this seems like a niche use case.

Eric

File: src/ray/raylet/worker_pool.cc (hunk 8 above)
Comment (kfstorm) [reply src/ray/raylet/worker_pool.cc:376]: @ericl I didn't count the test cases. I bet there are about a dozen test cases that depend on initial workers. Those test cases assert some conditions which depend on living workers, but they don't create any tasks. That's why they depend on initial workers. I can definitely fix these test cases. I've tried to fix them before. See [this PR](https://github.com/ray-project/ray/pull/7020).

**Comment by kfstorm** [comment]
Hi @ericl. @raulchen and I discussed this offline and we agree to not have a shared pool and disable num initial workers. As for the backward compatibility of the benchmark case, we agree that we need a trick to wait in `ray.init()`. We are still discussing capping total workers.

I've modified the PR to add [WIP] in the title and updated the description to include a checklist to track the issues we've discussed.

If you guys agree, I'll start coding below items:

* Disable num initial workers by default.
* If Raylet is spawned by `ray.init()`,
    * start `num_cpus` workers for the initial job.
    * Add a trick to wait for 1 worker in `ray.init()`.

**Comment by ericl** [comment]
That sounds good, thanks! Maybe we can wait for all N workers if we are
starting them all anyways .

**Comment by kfstorm** [comment]
@ericl I've updated the code. Please take a look again.

**Comment by ericl** [comment]
Thanks! Are you planning to add the overall pool size cap too here, or is it ready to review now?

**Comment by ericl** [comment]
One more thing, can we add a python unit test exercising multitenancy? This might require implementing env vars, not sure.

File: python/ray/worker.py [hunk 9]
```diff
@@ -588,6 +590,7 @@ def init(address=None,
         dashboard_port: The port to bind the dashboard server to. Defaults to
             8265.
         job_id: The ID of this job.
+        job_config (ray.job_config.JobConfig): The job configurations.
```
Comment (robertnishihara) [inline python/ray/worker.py:593]:
```suggestion
        job_config (ray.job_config.JobConfig): The job configuration.
```

**Comment by kfstorm** [comment]
> Thanks! Are you planning to add the overall pool size cap too here, or is it ready to review now?

I don't think we now have a conclusion and clear design about the strategy of capping yet. Let's discuss it more before implementing it.



> One more thing, can we add a python unit test exercising multitenancy? This might require implementing env vars, not sure.

Sure. I can port `MultiDriverTest.java` to Python.

File: python/ray/worker.py (hunk 9 above)
Comment (kfstorm) [reply python/ray/worker.py]: Done.

File: src/ray/raylet/worker_pool_test.cc [hunk 10]
```diff
@@ -346,5 +346,6 @@ TEST_F(WorkerPoolTest, StartWorkerWithDynamicOptionsCommand) {
 
 int main(int argc, char **argv) {
   ::testing::InitGoogleTest(&argc, argv);
+  RayConfig::instance().initialize({{"enable_multi_tenancy", "true"}});
```
Comment (stephanie-wang) [inline src/ray/raylet/worker_pool_test.cc:349]: Can we add some unit tests for multitenancy? Maybe:
1. Make sure PopWorker returns the right type of worker when there are multiple jobs.
2. Making sure the callback for creating a driver only gets called once all the initial workers have started (assuming we move that code into the WorkerPool and multitenancy is disabled).

Also, can we also add cases/run the suite for enable_multi_tenancy=false? Especially considering this is the default.

**Review by stephanie-wang — COMMENTED** [review]
Main comment is to try and move more logic out of NodeManager and into WorkerPool.

**Comment by ericl** [comment]
@kfstorm splitting off capping in another PR sounds good, though we probably can't enable multitenancy by default until that is done.

With regard to the design, why not something simple like preserving the existing capping of num_cpus?

File: src/ray/raylet/node_manager.cc [hunk 11]
```diff
@@ -1223,6 +1252,23 @@ void NodeManager::ProcessRegisterClientRequestMessage(
...
+    } else {
+      // This is a workaround to finish driver registration after all initial workers are
+      // registered to Raylet if and only if Raylet is started by a Python driver and the
+      // job config is not set in `ray.init(...)`.
+      if (RayConfig::instance().enable_multi_tenancy()) {
+        RAY_CHECK(!worker->GetAssignedJobId().IsNil());
+        if (first_job_ == worker->GetAssignedJobId() &&
+            worker->GetLanguage() == Language::PYTHON) {
+          ++first_job_registered_python_worker_count_;
+          if (first_job_send_register_client_reply_to_driver_ != nullptr &&
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply src/ray/raylet/node_manager.cc]: Good suggestion. Done.

File: src/ray/raylet/node_manager.cc [hunk 12]
```diff
@@ -1232,38 +1278,50 @@ void NodeManager::ProcessRegisterClientRequestMessage(
...
+          }
+        });
+  };
+
+  if (message->is_worker() || !RayConfig().instance().enable_multi_tenancy() ||
+      !first_job_.IsNil() || first_job_driver_wait_num_python_workers_ == 0 ||
+      first_job_send_register_client_reply_to_driver_ != nullptr) {
+    send_reply();
+  } else {
+    first_job_send_register_client_reply_to_driver_ = std::move(send_reply);
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply src/ray/raylet/node_manager.cc]: Done.

File: src/ray/raylet/node_manager.cc [hunk 13]
```diff
@@ -1232,38 +1278,50 @@ void NodeManager::ProcessRegisterClientRequestMessage(
...
+    client->WriteMessageAsync(
+        static_cast<int64_t>(protocol::MessageType::RegisterClientReply), fbb.GetSize(),
+        fbb.GetBufferPointer(), [this, client](const ray::Status &status) {
+          if (!status.ok()) {
+            ProcessDisconnectClientMessage(client);
+          }
+        });
+  };
+
+  if (message->is_worker() || !RayConfig().instance().enable_multi_tenancy() ||
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply src/ray/raylet/node_manager.cc]: Done. But the code is now in `worker_pool.cc`.

File: src/ray/raylet/node_manager.h [hunk 14]
```diff
@@ -727,6 +729,26 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
...
+
+  /// The umber of initial Python workers to wait for the first job before the driver
+  /// receives RegisterClientReply.
+  int first_job_driver_wait_num_python_workers_;
+
+  /// The number of initial Python workers for the first job.
+  int num_initial_python_workers_for_first_job_;
+
+  /// This map tracks the latest infos of unfinished jobs.
+  absl::flat_hash_map<JobID, rpc::JobTableData> unfinished_jobs_;
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply src/ray/raylet/node_manager.h]: Done.

File: src/ray/raylet/worker_pool.cc [hunk 15]
```diff
@@ -153,8 +160,18 @@ uint32_t WorkerPool::Size(const Language &language) const {
...
-Process WorkerPool::StartWorkerProcess(const Language &language,
-                                       const std::vector<std::string> &dynamic_options) {
+Process WorkerPool::StartWorkerProcess(const Language &language, const JobID &job_id,
+                                       std::vector<std::string> dynamic_options) {
+  boost::optional<rpc::JobConfig> job_config;
+  if (RayConfig::instance().enable_multi_tenancy()) {
+    RAY_CHECK(!job_id.IsNil());
+    job_config = job_config_getter_(job_id);
+    if (!job_config) {
+      RAY_LOG(INFO) << "Job config of job " << job_id << " are not local yet.";
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply src/ray/raylet/worker_pool.cc:175]: Tasks will be dispatched again in `NodeManager::HandleJobStarted`. Sounds reasonable for you? Just like we have `maximum_startup_concurrency` and we also invoke `DispatchTasks` in `HandleWorkerAvailable`.

[the remark this answers is not in the document]
Comment (kfstorm) [reply src/ray/raylet/worker_pool.cc:175]: I've added a comment here.

File: src/ray/raylet/worker_pool.cc [hunk 16]
```diff
@@ -240,13 +272,20 @@ Process WorkerPool::StartWorkerProcess(const Language &language,
         << " placeholder is not found in worker command.";
   }
 
+  // TODO(kfstorm): Set up environment variables in a later PR.
   Process proc = StartProcess(worker_command_args);
+  if (RayConfig::instance().enable_multi_tenancy()) {
+    // If the pid is reused between processes, the old process must have exited.
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply src/ray/raylet/worker_pool.cc:285]: `ProcessDisconnectClientMessage` will be invoked if the worker process is dead because the RPC connection is broken. But it has nothing to do with this comment. The comment is talking about what if a newly created worker process has the PID which is also used by a historical worker process probably a long time ago. In this case, the old worker process must have already been dead. So it's safe to update the value of the <pid, job id> map.

File: src/ray/raylet/worker_pool.cc [hunk 17]
```diff
@@ -53,19 +53,19 @@ namespace ray {
...
 
-/// A constructor that initializes a worker pool with num_workers workers for
-/// each language.
-WorkerPool::WorkerPool(boost::asio::io_service &io_service, int num_workers,
-                       int maximum_startup_concurrency, int min_worker_port,
-                       int max_worker_port, std::shared_ptr<gcs::GcsClient> gcs_client,
-                       const WorkerCommandMap &worker_commands,
-                       const std::unordered_map<std::string, std::string> &raylet_config,
-                       std::function<void()> starting_worker_timeout_callback)
+WorkerPool::WorkerPool(
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply src/ray/raylet/worker_pool.cc]: Some code blocks are not related to job ID. So it's not easy to simplify all the if-else blocks. But as I said before, the new code is compatible with existing jobs except for some config changes (the ones now in job configs). As long as no two jobs are deployed to the same cluster, the new code should work just fine. So my suggestion actually is to remove the flag.

@ericl You may have concerns about the worker-capping feature. The new code has no design change in resource management. CPU is still a node-level resource, not a per-node-per-job resource. Let's say that the `num_cpus` is set to 8 on a node, and we have two jobs on this node each has already occupied 3 and 4 CPUs, then there's no way to schedule a new actor/task which requires 2 CPUs, for any job.

We are actively testing the multi-tenancy feature with different kinds of jobs on one cluster. And this is not the final PR about multi-tenancy. We definitely shouldn't try to adopt this feature (by running multiple jobs into one cluster) unless it's fully production-ready. Maybe I should rename the `job_config` parameter of `ray.init()` to `_job_config` for now, just like `_internal_config`.

**Comment by kfstorm** [comment]
@ericl See my comment [here](https://github.com/ray-project/ray/pull/9500#discussion_r461444066) about preserving the existing capping of num_cpus.

File: src/ray/raylet/worker_pool_test.cc (hunk 10 above)
Comment (kfstorm) [reply src/ray/raylet/worker_pool_test.cc:340]: 1. Yeah, I can add that.
2. I assume you were referring to the `send_reply_callback` of the `RegisterDriver` method. If multi-tenancy is disabled, the callback will be invoked immediately and the old logic of spawning initial workers in the constructor `WorkerPool` for every job is kept.

I'll try to make it testable for both true and false of the `enable_multi_tenancy` flag.

File: src/ray/raylet/worker_pool.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply src/ray/raylet/worker_pool.cc:175]: Ah that makes sense. Thanks!

File: src/ray/raylet/worker_pool_test.cc (hunk 10 above)
Comment (kfstorm) [reply src/ray/raylet/worker_pool_test.cc:340]: @stephanie-wang If multi-tenancy is disabled, the driver's `ray.init()` will not wait for the initial workers to be started. But for other reasons, it looks like it waited. See [here](https://github.com/ray-project/ray/pull/9500#issuecomment-659791940) for the explanation. Do you want me to change the behavior to always wait for initial workers no matter whether multi-tenancy is enabled or not?

Comment (kfstorm) [reply src/ray/raylet/worker_pool_test.cc]: I've added a test case for 1) and make worker pool test available for both enable_multi_tenancy=true and enable_multi_tenancy=false.

Comment (stephanie-wang) [reply src/ray/raylet/worker_pool_test.cc]: That sounds good to me! It's fine if we don't cover case 2.

File: python/ray/tests/test_multi_tenancy.py [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+ray.shutdown()
+    """
+
+    f = tempfile.NamedTemporaryFile(suffix=".py")
+    f.write(driver_code.encode("ascii"))
+    f.flush()
+
+    driver_count = 10
+    cmd = [sys.executable, f.name, info["redis_address"]]
+    processes = [
```
Comment (rkooo567) [inline python/ray/tests/test_multi_tenancy.py:89]: Why don't you just use `run_string_as_driver_nonblocking`?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 19]
```diff
@@ -125,7 +125,9 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
       worker->SetAllocatedInstances(allocated_instances);
     }
     worker->AssignTaskId(spec.TaskId());
-    worker->AssignJobId(spec.JobId());
+    if (!RayConfig::instance().enable_multi_tenancy()) {
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:128]: Why do we need this if? It looks like we already check this inside `AssignJobId` method.

File: src/ray/raylet/worker_pool.cc [hunk 20]
```diff
@@ -241,13 +279,20 @@ Process WorkerPool::StartWorkerProcess(const Language &language,
...
+    RAY_LOG(INFO) << "Worker process " << proc.GetId() << " is bound to job " << job_id;
+    state.worker_pids_to_assigned_jobs[proc.GetId()] = job_id;
+  }
   RAY_LOG(DEBUG) << "Started worker process of " << workers_to_start
                  << " worker(s) with pid " << proc.GetId();
   MonitorStartingWorkerProcess(proc, language);
   state.starting_worker_processes.emplace(proc, workers_to_start);
   return proc;
-}
+}  // namespace raylet
```
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:295]: Maybe mistake?

File: src/ray/raylet/worker_pool.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (ericl) [reply src/ray/raylet/worker_pool.cc]: @kfstorm that scenario doesn't make sense, workers should have nothing to do with the resource management. If we need workers to launch a task, we should create new ones, and kill idle workers to keep the total excess usage under control.

[the remark this answers is not in the document]
Comment (kfstorm) [reply src/ray/raylet/worker_pool.cc]: Ideally, we can do that, and we should do that. But for now, as long as we don't run multiple jobs in the same cluster, it should be OK to turn multi-tenancy feature on by default. That's why I suggest that we can actually remove the feature flag since @stephanie-wang is worried about the complexity of if-elses.

File: src/ray/raylet/worker_pool_test.cc (hunk 10 above)
Comment (kfstorm) [reply src/ray/raylet/worker_pool_test.cc]: BTW, case 2 with multi-tenancy on is covered by `test_initial_workers` in `test_multi_tenancy.py`.

File: python/ray/tests/test_advanced_2.py [hunk 21]
```diff
@@ -448,6 +448,9 @@ def test_custom_resources(ray_start_cluster):
         num_cpus=1, resources={"CustomResource": 1})
     ray.init(address=cluster.address)
 
+    # Wait for workers to be ready to execute tasks
+    time.sleep(3)
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply python/ray/tests/test_advanced_2.py:452]: Is the issue resolved? Previously it's hard to pass this test. I'll revert it since it's not related to this PR.

File: python/ray/tests/test_multi_tenancy.py (hunk 18 above)
Comment (kfstorm) [reply python/ray/tests/test_multi_tenancy.py:89]: Thanks for the heads-up! I didn't know this helper method.

File: src/ray/raylet/scheduling/cluster_task_manager.cc (hunk 19 above)
Comment (kfstorm) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:128]: Sure. I'll revert this. The check inside `AssignJobId` is for safety concern. Actually, the `AssignJobId` will be removed once we can pass the job ID to worker processes via an env variable, in a later PR.

File: src/ray/raylet/worker_pool.cc (hunk 20 above)
Comment (kfstorm) [reply src/ray/raylet/worker_pool.cc:295]: Thanks!

File: src/ray/raylet/worker_pool.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply src/ray/raylet/worker_pool.cc]: I'm not really worried about the if-elses; just wanted to see if there was a way to avoid them.

**Comment by kfstorm** [comment]
@eric @robertnishihara @stephanie-wang @rkooo567 @raulchen Any more comments?

**Review by ericl — APPROVED** [review]
LGTM to merge as an experimental feature

**Comment by edoakes** [comment]
@kfstorm I just want to bring to your attention that we'll probably want to be able to specify these for different workers/actors within a single job. We should not make any assumptions here that preclude doing that in the future.

**Comment by kfstorm** [comment]
You mean the environment variables? You can always implement actor level options later. For example, we already have JVM options available as an actor creation option in Java.
________________________________

**Comment by ericl** [comment]
Merge conflict
